sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -968,18 +1146,55 @@ public void validate(PCollection<T> input) {
 
     @Override
     public WriteResult expand(PCollection<T> input) {
-      PCollection<KV<TableDestination, TableRow>> rowsWithDestination =
-          input.apply("PrepareWrite", new PrepareWrite<T>(
-              getTableFunction(), getFormatFunction()))
-              .setCoder(KvCoder.of(TableDestinationCoder.of(), TableRowJsonCoder.of()));
+      DynamicDestinations<T, ?> dynamicDestinations = getDynamicDestinations();
+      if (dynamicDestinations == null) {
+        if (getJsonTableRef() != null) {
+          dynamicDestinations = ConstantTableDestinations.fromJsonTableRef(
+              getJsonTableRef(), getTableDescription());
+        } else if (getTableFunction() != null) {
+          dynamicDestinations = new TableFunctionDestinations(getTableFunction());
+        }
+
+        // Wrap with a DynamicDestinations class that will provide a schema. There might be no
+        // schema provided if the create disposition is CREATE_NEVER.
+        if (getJsonSchema() != null) {
+          dynamicDestinations = new ConstantSchemaDestinations(
+              dynamicDestinations, getJsonSchema());
+        } else if (getSchemaFromView() != null) {
+          dynamicDestinations = new SchemaFromViewDestinations(
+              dynamicDestinations, getSchemaFromView());
+        }
+      }
+      return expandTyped(input, dynamicDestinations);
+    }
 
+    private <DestinationT> WriteResult expandTyped(
+        PCollection<T> input, DynamicDestinations<T, DestinationT> dynamicDestinations) {
+      Coder<DestinationT> destinationCoder = dynamicDestinations.getDestinationCoder();
+      if (destinationCoder == null) {
+        try {
+          // If dynamicDestinations doesn't provide a coder, try to find it in the coder registry.
+          destinationCoder = input.getPipeline().getCoderRegistry().getDefaultCoder(
+              new TypeDescriptor<DestinationT>() {
```

Comment:
This isn't gonna work, `DestinationT` is erased at this point because it is not a type parameter of a base class of the current class. You need to get the type parameter value from the `DynamicDestinations` object. Is there a test for this?